### PR TITLE
proper gradient clipping

### PR DESCRIPTION
**Pull request by justheuristic**
Currently, each peer runs gradient clipping separately for each micro-batch of gradients.

For proper training, the gradient used in opt.step *should* be **clip(average(microbatch_gradients))**, but currently it is **average(map(clip(microbatch_gradient)))**. Since the clipping norm is 1, this will introduce a significant bias in gradients, downplaying the "weight" of samples with high-magnitude gradients.

This pull-request introduces a more "proper" variant of gradient clipping, wherein peers only clip gradient once per step after computing the average of non-cliped gradients.

### CODE REVIEW

**Comment by justheuristic** [comment]
Also: while we're figuring out how to wandb it properly, here's a neat little hack:
image

Since LR goes up linearly, we can use it as a proxy measure of time:
image


__Note: this will only work for the first few iterations while LR is in the warmup phase__

**Comment by justheuristic** [comment]
Here's my current training run:
https://wandb.ai/justheuristic/swarm

I was using target_batch_size=8192 and per-worker batch_size=32 because our 1080Tis are not as fast as your 3090s and I wanna see convergence in my lifetime)

File: wikitext.py [hunk 1]
```diff
@@ -605,7 +611,7 @@ def on_validation_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModul
         log_every_n_steps=1,
         max_epochs=args.max_epochs,
         precision=16,
-        gradient_clip_val=1,
+        gradient_clip_val=99999999999999999999999999,
```
Comment (SeanNaren) [inline wikitext.py:614]:
```suggestion
```
